src/avx512bw/avx512bw-from-valid-utf8.inl.cpp

```diff
@@ -0,0 +1,339 @@
+        |0000.0001|10aa.aabb|bbbb.cccc|cc??.????| 3-byte char
+        |0000.0001|0aaa.aabb|bbbb.????|????.????| 2-byte char
+        |0000.000a|aaaa.aa??|????.????|????.????| ASCII char */
+    values = _mm512_madd_epi16(values, v_0001_1000);
+
+    /* 4. Shift left the values by variable amounts to reset highest UTF-8 bits
+        |aaab.bbbb|bccc.cccd|dddd.d000|0000.0000| 4-byte char -- by 11
+        |aaaa.bbbb|bbcc.cccc|????.??00|0000.0000| 3-byte char -- by 10
+        |aaaa.abbb|bbb?.????|????.???0|0000.0000| 2-byte char -- by 9
+        |aaaa.aaa?|????.????|????.????|?000.0000| ASCII char -- by 7 */
+    {
+        /** pshufb
+
+        continuation = 0
+        ascii    = 7
+        _2_bytes = 9
+        _3_bytes = 10
+        _4_bytes = 11
+
+        shift_left_v3 = 4 * [
+            ascii, # 0000
+            ascii, # 0001
+            ascii, # 0010
+            ascii, # 0011
+            ascii, # 0100
+            ascii, # 0101
+            ascii, # 0110
+            ascii, # 0111
+            continuation, # 1000
+            continuation, # 1001
+            continuation, # 1010
+            continuation, # 1011
+            _2_bytes, # 1100
+            _2_bytes, # 1101
+            _3_bytes, # 1110
+            _4_bytes, # 1111
+        ] */
+        const __m512i shift_left_v3 = _mm512_setr_epi64(
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000
+        );
+
+        const __m512i shift = _mm512_shuffle_epi8(shift_left_v3, char_class);
+        values = _mm512_sllv_epi32(values, shift);
+    }
+
+    /* 5. Shift right the values by variable amounts to reset lowest bits
+        |0000.0000|000a.aabb|bbbb.cccc|ccdd.dddd| 4-byte char -- by 11
+        |0000.0000|0000.0000|aaaa.bbbb|bbcc.cccc| 3-byte char -- by 16
+        |0000.0000|0000.0000|0000.0aaa|aabb.bbbb| 2-byte char -- by 21
+        |0000.0000|0000.0000|0000.0000|0aaa.aaaa| ASCII char -- by 25 */
+    {
+        // 4 * [25, 25, 25, 25, 25, 25, 25, 25, 0, 0, 0, 0, 21, 21, 16, 11]
+        const __m512i shift_right = _mm512_setr_epi64(
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000
+        );
+
+        const __m512i shift = _mm512_shuffle_epi8(shift_right, char_class);
+        values = _mm512_srlv_epi32(values, shift);
+    }
+
+    return values;
+}
+
+
+// See: http://0x80.pl/notesen/2021-12-22-test-and-clear-bit.html
+bool test_and_clear_bit(uint32_t& val, int bitpos) {
+    const uint32_t bitmask = uint32_t(1) << bitpos;
+    const uint32_t old = val;
+
+    val &= ~bitmask;
+
+    return val < old;
+}
```

Comment:
Fun fact: this function compiles to something efficient under GCC and clang:
https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEEgDMpAAOqAqETgwe3r7%2ByanpAiFhkSwxcVyJdpgOGUIETMQEWT5%2BAVU1AnUNBEUR0bEJtvWNzTltwz2hfaUDFQCUtqhexMjsHAD06wDUACoAnkmYW3tLxFtoWFsIsZikWyRbtKhM6FuGW5iqrEn0AHQmGgAguZ4qFkN5LiZ4m4FAR8IJfggodgAcCgVFUJ4tgRMLCAPqGdB48GYBp4qKECBeUIEeJmPEEcwANi2ADcxHcaVsKQQUgo5lsTAB2KxArbi8VoBiwrbUwR0hncwgsJgKADWgviABFZTSFQQIFwBVC3CalbzUlDRYCJZKBDK5bT6QR7rRXlCdezaFbUajbWyxIKzEyPVsAH48lXqn1Av224iYAjLBgB2iatyu9Axm3CrW%2BsUcBa0TgAVl4fg4WlIqE4pss1i2ClOqyD8R4pAImkLCzVIHi8V%2B/aHw5HTP0nEk5a71c4vAUIA0Ha7CzgsCQaBYSTosXIlA3W/ocWAXC4ARotBxxHnECi04pzGIe047Y3bEEAHkGLQn5XeFgVUY4i/qQ%2BAJjUrK4tOnzVF4OLPrwNKYMWwG0HgUTEA0eweFg04EMQeAsPBCxUAYwAKAAangmAAO7vocFbtvwggiGI7BSDIgiKCo6jAboXD6IYxjWNY%2BhofOkALKgSSOPanAALTvvEWxySqKxItqzCOBBc5IdUMl%2BBArijH4/HBFMJRlHofIFJkngtFZ%2BT6b0FmzLYukdAwXQjHZOT8e0%2BleZMxT9OUQzdMZeiwt0zkhRICxNssqx6HhmBrDwRallOwE1hwqgABxMnJTKSFswDIMgWwnr8ZhbBAdZWJYdy4IQDwgvxWweJu25nG1cy8J2v5zD2fYDiOY1DmOyGTqQFZVjlc4Lkug2kKuiAoKgXWHruED7t1KAGEYFQaIu56Xtet7AfemHwaQr6MAQn7ftO/6CUBVagXpeAQfOwHQcgsFpQhghIdOqHoZh2FrFWeEEURfCkRRVG0fRN1McIojiOxaNcWo066AEB1CfWjVg%2BJECSdJGQ/QpSkqQ0yDqVqmlfZgOn2PpLgMO4PkmYEXMxTMoXWfpEX8cLGQC5ZfnuQFEyi257O1BMkuuVF3nZLzatBdMUvxc2bEdgmgMZRwZYzdOOX5YVxXnITlUDhovwaLV9UiVszVED1ZhtncnUHrErZmH1S1aENpC9v2g7jWN44cNNs28PNtiLQNocm2YWVzbOIfdqQEFXhkICSEAA%3D%3D%3D

The assembly output includes btr.

Reply:
Yes, it does, but the following code emits `cmp` anyway. I noted this here: http://0x80.pl/notesen/2021-12-22-test-and-clear-bit.html#alternative